### PR TITLE
introduce CustomResourceDefinitions monitor target

### PR DESCRIPTION
**Pull request by zouyee**
**What this PR does / why we need it**:
If CRD service could be monitored , it could detect the health status of crd to identify problems.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #https://github.com/kubernetes/kube-state-metrics/issues/514

/cc @brancz @andyxning

### CODE REVIEW

File: Documentation/customresourcedefinition-metrics.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Stateful Set Metrics
```
Comment (mxinden) [inline Documentation/customresourcedefinition-metrics.md]: `s/Stateful Set/CustomResourceDefinition Metrics/`

File: pkg/collectors/customresourcedefinitions.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
+/*
```
Comment (mxinden) [inline pkg/collectors/customresourcedefinitions.go]: All the other files are called by the singular name of the object. What are your thoughts?

Comment (brancz) [reply pkg/collectors/customresourcedefinitions.go]: Agreed, let's be consistent.

File: Documentation/README.md [hunk 3]
```diff
@@ -58,4 +58,5 @@ Per group of metrics there is one file for each metrics. See each file for speci
 * [Horizontal Pod Autoscaler Metrics](horizontalpodautoscaler-metrics.md)
 * [Endpoint Metrics](endpoint-metrics.md)
 * [Secret Metrics](secret-metrics.md)
-* [ConfigMap Metrics](configmap-metrics.md)
\ No newline at end of file
+* [ConfigMap Metrics](configmap-metrics.md)
+* [CustomResourceDefinitions Metrics](customresourcedefinition-metrics.md)
```
Comment (brancz) [inline Documentation/README.md]: nit: let's add a new line here so the diff is more precise in the future

File: Documentation/customresourcedefinition-metrics.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+| Metric name| Metric type | Labels/tags | Status |
+| ---------- | ----------- | ----------- | ----------- |
+| kube_customresourcedefinition_created | Gauge | `customresourcedefinition`=&lt;customresourcedefinition-name&gt; | EXPERIMENTAL |
+| kube_customresourcedefinition_labels | Gauge | `customresourcedefinition`=&lt;customresourcedefinition-name&gt; <br> `label_CRD_LABEL`=&lt;CRD_LABEL&gt;  | EXPERIMENTAL |
```
Comment (brancz) [inline Documentation/customresourcedefinition-metrics.md]: is this actually useful? I've never seen anyone label a CustomResourceDefinition

Comment (zouyee) [reply Documentation/customresourcedefinition-metrics.md]: https://github.com/kubernetes/cluster-registry/blob/master/cluster-registry-crd.yaml#L7

Comment (brancz) [reply Documentation/customresourcedefinition-metrics.md]: fair :+1:

**Comment by brancz** [comment]
I think generally this is useful, but keep in mind that we are currently in feature freeze and I think I would prefer to have https://github.com/kubernetes/kube-state-metrics/pull/513 land first, which has some architectural changes affecting this pull request. I'd say let's get #513 in once feature freeze is lifted and then work towards getting this down as well.

**Comment by zouyee** [comment]
got it！@brancz

**Comment by zouyee** [comment]
ping @brancz

**Comment by brancz** [comment]
@mxinden can you make sure that the changes here architecturally don't interfere with the performance optimizations you are working on?

**Comment by zouyee** [comment]
@brancz @mxinden @andyxning  PTAL

File: pkg/collectors/daemonset.go [hunk 5]
```diff
@@ -99,11 +100,21 @@ func (l DaemonSetLister) List() ([]v1beta1.DaemonSet, error) {
...
 }
 
-func RegisterDaemonSetCollector(registry prometheus.Registerer, informerFactories []informers.SharedInformerFactory, opts *options.Options) {
+func RegisterDaemonSetCollector(registry prometheus.Registerer, informerFactories map[string][]interface{}, opts *options.Options) {
 
 	infs := SharedInformerList{}
-	for _, f := range informerFactories {
-		infs = append(infs, f.Extensions().V1beta1().DaemonSets().Informer().(cache.SharedInformer))
+	gvr := schema.GroupVersionResource{
+		Group:    "extensions",
```
Comment (mxinden) [inline pkg/collectors/daemonset.go]: Passing these as a struct of strings makes it easier to generalize in the long run, but does not catch typos during compile time. Not a blocker, just something to keep in mind.

Comment (zouyee) [reply pkg/collectors/daemonset.go]: 👌

**Comment by brancz** [comment]
this needs a rebase/rewrite

**Comment by zouyee** [comment]
I will fix it in this weekly.

**Comment by jinxingwang** [comment]
@zouyee HI. why this PR is closed?
We are a team using sparkapplication heavily.
adding metrics for CustomResourceDefinitions will be super useful to us.

**Comment by hairyhenderson** [comment]
@zouyee any plans to re-open this? It'd be really useful to have metrics around CRDs.

**Comment by rdubya16** [comment]
@zouyee What happened to this? Would have be useful to us as well.

**Pull request closed without merging**